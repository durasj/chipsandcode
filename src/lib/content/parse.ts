import Markdoc, { type Config, type RenderableTreeNode } from '@markdoc/markdoc';
import { parse } from 'yaml';
import type { Content } from './types';

const config = {
  tags: {
    math: {
      render: 'Math',
      description: 'Math KaTeX expression',
      inline: true,
      selfClosing: true,
      attributes: {
        expr: {
          type: String,
        },
      },
    },
    EmbeddedHardwareIDE: {
      render: 'EmbeddedHardwareIDE',
      description: 'Embedded Hardware IDE',
      selfClosing: true,
      attributes: {
        id: {
          type: String,
        },
      },
    },
    EmbeddedStartingFiles: {
      render: 'EmbeddedStartingFiles',
      description: 'Embedded Starting Files',
      selfClosing: true,
      attributes: {
        id: {
          type: String,
        },
        name: {
          type: String,
        },
      },
    },
    EmbeddedVideo: {
      render: 'EmbeddedVideo',
      description: 'Embedded YouTube Video',
      selfClosing: true,
      attributes: {
        id: {
          type: String,
        },
      },
    },
  },
} satisfies Config;

const parseMarkdown = (doc: string) => {
  // Primitive way to support inline $math$ expressions
  // Markdoc doesn't support custom syntax like this
  const expandedDoc = doc.replaceAll(/\$(.+?)\$/g, '{% math expr="$1" /%}');

  const tokenizer = new Markdoc.Tokenizer({
    allowIndentation: true,
    allowComments: true,
    linkify: true,
    typographer: true,
  });
  const tokens = tokenizer.tokenize(expandedDoc);
  const ast = Markdoc.parse(tokens);

  const meta: unknown = ast.attributes.frontmatter ? parse(ast.attributes.frontmatter) : undefined;

  const content = Markdoc.transform(ast, { ...config, variables: { meta } });

  // Transform headings so that we start with h2 - there is only one h1 and it is already on the page
  const headingMap = {
    h1: 'h2',
    h2: 'h3',
    h3: 'h4',
    h4: 'h5',
    h5: 'h6',
    h6: 'h7',
  };
  const shiftHeading = (node: RenderableTreeNode): RenderableTreeNode => {
    if (!node || !(typeof node === 'object') || Array.isArray(node)) return node;

    const mappedHeading =
      typeof node.name === 'string' && node.name in headingMap
        ? headingMap[node.name as keyof typeof headingMap]
        : undefined;

    const children = Array.isArray(node.children) ? node.children.map(shiftHeading) : node.children;

    return mappedHeading
      ? ({ ...node, name: mappedHeading, children } as RenderableTreeNode)
      : ({ ...node, children } as RenderableTreeNode);
  };

  const processedContent = shiftHeading(content);

  return { meta, content: processedContent } as Content;
};

export default parseMarkdown;
