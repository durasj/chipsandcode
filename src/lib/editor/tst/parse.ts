import { Parser, Grammar } from 'nearley';

import grammar from './grammar.ne';
import type { Root } from './tree';

/**
 * Parses provided test script source code to AST
 *
 * @param input test script source code
 */
function parse(input: string): Root | undefined {
  const parser = new Parser(Grammar.fromCompiled(grammar));
  parser.feed(input);

  if (parser.results.length > 1) {
    // TODO: Log to Sentry once (if) it gets integrated
    console.warn('Ambiguous parser result!', parser.results);
  }

  return parser.results[0];
}

export default parse;
