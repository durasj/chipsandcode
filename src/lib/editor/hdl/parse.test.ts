import { describe, expect, it } from 'vitest';
import { extname, basename } from 'node:path';
import parse from './parse';

const emptyChip = `CHIP And {}`;
const chipWithoutName = `CHIP { IN a; }`;
const minimalChip = `CHIP Foo { IN a; }`;
const chipWithEmptyParts = `CHIP Foo { PARTS: }`;

describe('HDL Parser', () => {
  it('Parses empty HDL', () => {
    const output = parse('');

    expect(output).toBeUndefined();
  });

  it('Does not parse an empty CHIP', () => {
    expect(() => parse(emptyChip)).toThrowError(
      /Syntax error at line 1 col 11:\n\n1 {2}CHIP And {}/m,
    );
  });

  it('Does not parse CHIP without name', () => {
    expect(() => parse(chipWithoutName)).toThrowError(
      /Syntax error at line 1 col 6:\n\n1 {2}CHIP { IN a; }/m,
    );
  });

  it('Parses minimal CHIP with only pin definition', () => {
    const ast = parse(minimalChip);

    expect(ast).toMatchSnapshot();
  });

  it('Does not parse empty PARTS', () => {
    expect(() => parse(chipWithEmptyParts)).toThrowError(
      /Syntax error at line 1 col 19:\n\n1 {2}CHIP Foo { PARTS: }/m,
    );
  });

  // Built-in HDL files
  const builtInHdls = import.meta.glob('./testing/builtin/*.hdl', { as: 'raw', eager: true });
  for (const [path, content] of Object.entries(builtInHdls)) {
    const name = basename(path, extname(path));

    it(`Parses Built-In ${name}`, () => {
      const ast = parse(content);

      expect(JSON.stringify(ast, null, 2)).toMatchFileSnapshot(`./testing/builtin/${name}.hdl.ast`);
    });
  }

  // Project 1 HDL files
  const project1Hdls = import.meta.glob('./testing/project1/*.hdl', { as: 'raw', eager: true });
  for (const [path, content] of Object.entries(project1Hdls)) {
    const name = basename(path, extname(path));

    it(`Parses Project 1 ${name}`, () => {
      const ast = parse(content);

      expect(JSON.stringify(ast, null, 2)).toMatchFileSnapshot(
        `./testing/project1/${name}.hdl.ast`,
      );
    });
  }
});
