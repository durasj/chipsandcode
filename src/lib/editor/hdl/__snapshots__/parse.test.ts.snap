// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HDL Parser > Parses minimal CHIP with only pin definition 1`] = `
[
  {
    "body": [
      {
        "pins": [
          [
            {
              "col": 15,
              "line": 1,
              "lineBreaks": 0,
              "offset": 14,
              "value": "a",
            },
          ],
        ],
        "type": "input",
      },
    ],
    "name": {
      "col": 6,
      "line": 1,
      "lineBreaks": 0,
      "offset": 5,
      "value": "Foo",
    },
    "type": "chip",
  },
]
`;
